### PR TITLE
Peer worker file transfers

### PR DESCRIPTION
**Pull request by piercecunneen**

### CODE REVIEW

**Comment by dthain** [comment]
@piercecunneen please check the build failures here:
http://ccl.cse.nd.edu/software/autobuild

File: work_queue/src/work_queue.c [hunk 1]
```diff
@@ -955,13 +996,18 @@ static void add_worker(struct work_queue *q)
...
+	
+	w->worker_link_port = -1;
+	
 	w->last_update_msg_time = w->start_time;
 
 	w->resources = work_queue_resources_create();
 
 	w->workerid = NULL;
 
+	w->is_fetching = 0;
```
Comment (piercecunneen) [inline work_queue/src/work_queue.c]: this member variable is not necessary, as when a worker is not involved in a fetch then w->ongoing_fetch == NULL. Will remove in the near future

**Comment by piercecunneen** [comment]
This is very much a work in progress. Things that still need to be considered.

1. Stats reporting
**1.a** What stats do we want to keep track of (e.g. bytes fetched, bytes provided, total fetching time, etc)? 
**1.b** What stats should individual workers keep track of vs what stats should the master hold in aggregate 

2. Handling failures
**2.a** Need to add a new failure type (in addition to worker and app failure) that describes a fetch failure properly. When a fetch failure occurs, we need to resend the task that the fetch occurred during. 
**2.b** When should we stop fetching and default to the traditional pattern of the master sending all files? 
**2.c** When should we prevent a specific worker from taking part in a fetch operation? 
**2.d** Differentiating between the failure occurring on the providing worker vs. occurring on the fetching worker.

**Comment by dthain** [comment]
@piercecunneen please stop by and talk to @btovar about this.  Sketch out what you are trying to do and get his feedback on some of the details.

For failure handling, I suggest something simple: "provide" should expire after some large fixed timeout.  (30 mins?)  "fetch" should fail after some timeout proportional to the file size.  Then, at the master side, have a threshhold for fetch failures per worker (1?) and failures overall (3?).  If you exceed fetch failures for a given worker, then only use the old method on that worker.  If you exceed failures overall, then only use the old method everywhere.

File: work_queue/src/work_queue.c [hunk 2]
```diff
@@ -210,6 +212,11 @@ struct work_queue_worker {
 	struct work_queue_stats     *stats;
 	struct work_queue_resources *resources;
 
+	int worker_link_port;
+
+  int is_fetching;
```
Comment (btovar) [inline work_queue/src/work_queue.c]: (I see you have a comment related to this.)

File: work_queue/src/work_queue.c [hunk 3]
```diff
@@ -314,6 +328,11 @@ static struct list *work_queue_task_file_list_clone(struct list *list);
 /** Write master's resources to resource summary file and close the file **/
 void work_queue_disable_monitoring(struct work_queue *q);
 
+
+static struct work_queue_worker *find_worker_by_random(struct work_queue *q, struct work_queue_task *t);
+
+static struct work_queue_worker *find_worker_to_fetch_from(struct work_queue *q, const char *filename, struct work_queue_worker *wk, int taskid);
```
Comment (btovar) [inline work_queue/src/work_queue.c]: *find_worker_to_fetch_file_from

File: work_queue/src/work_queue.c [hunk 4]
```diff
@@ -907,6 +935,20 @@ static int release_worker(struct work_queue *q, struct work_queue_worker *w)
...
 
+static int remove_fetch(struct work_queue *q, struct work_queue_worker *w) {
+	struct fetch *f;
+	list_first_item(q->ongoing_fetches);
+	while ((f=list_next_item(q->ongoing_fetches))) {
+		if (w == f->to_worker || w== f->from_worker) {
+			f->to_worker->ongoing_fetch = NULL;
+			f->from_worker->ongoing_fetch = NULL;
+			list_remove(q->ongoing_fetches, f);
+			return 1;
```
Comment (btovar) [inline work_queue/src/work_queue.c]: memory leaks. Need to free(f) and perhaps whatever ->ongoing_fetch is pointing to.

File: work_queue/src/work_queue.c [hunk 5]
```diff
@@ -955,13 +996,18 @@ static void add_worker(struct work_queue *q)
 	w->current_tasks_boxes = itable_create(0);
 	w->finished_tasks = 0;
 	w->start_time = timestamp_get();
-
+	
+	w->worker_link_port = -1;
```
Comment (btovar) [inline work_queue/src/work_queue.c]: worker_fetch_file_port

File: work_queue/src/work_queue.c [hunk 6]
```diff
@@ -2599,7 +2645,16 @@ static int send_file( struct work_queue *q, struct work_queue_worker *w, struct
 	if(!length) {
 		length = local_info.st_size;
 	}
-
+	if (flags & WORK_QUEUE_REMOTE && w->ongoing_fetch == NULL) {
```
Comment (btovar) [inline work_queue/src/work_queue.c]: if (flags & WORK_QUEUE_REMOTE && !w->ongoing_fetch) {

File: work_queue/src/work_queue.c [hunk 7]
```diff
@@ -2599,7 +2645,16 @@ static int send_file( struct work_queue *q, struct work_queue_worker *w, struct
...
+		struct work_queue_worker *worker;
+		worker = find_worker_to_fetch_from(q, remotename, w, t->taskid);
+		if (worker) {
+			printf("sending fetch %s from %s to %s at port %d\n", remotename, worker->hostname, w->hostname, worker->worker_link_port);
```
Comment (btovar) [inline work_queue/src/work_queue.c]: Instead of printf(  use debug(D_WQ,

File: work_queue/src/work_queue.c [hunk 8]
```diff
@@ -3270,6 +3330,49 @@ static struct work_queue_worker *find_worker_by_random(struct work_queue *q, str
 	return w;
 }
 
+int is_valid_fetch(struct work_queue *q, struct fetch *Fetch) {
```
Comment (btovar) [inline work_queue/src/work_queue.c]: Once the struct is renamed as above, make Fetch to fetch.

File: work_queue/src/work_queue.c [hunk 9]
```diff
@@ -3270,6 +3330,49 @@ static struct work_queue_worker *find_worker_by_random(struct work_queue *q, str
...
+	while(hash_table_nextkey(q->worker_table, &key, (void**)&w)) {
+		remote_info = hash_table_lookup(w->current_files, filename);
+		struct fetch *potential_fetch = malloc(sizeof(struct fetch));
+		potential_fetch->from_worker = w;
+		potential_fetch->to_worker = wk;
+		potential_fetch->filename = malloc(strlen(filename) + 1);
+		potential_fetch->taskid = taskid;
+		strncpy(potential_fetch->filename,filename, strlen(filename));
+		potential_fetch->filename[strlen(filename)] = '\0';
+		if (remote_info && strcmp(wk->hostname, w->hostname) && is_valid_fetch(q,potential_fetch)) {
```
Comment (btovar) [inline work_queue/src/work_queue.c]: and the worker is not in q->workers_with_available_results

File: work_queue/src/work_queue_worker.c [hunk 10]
```diff
@@ -207,6 +209,12 @@ static int total_tasks_executed = 0;
 static const char *project_regex = 0;
 static int released_by_master = 0;
 
+static struct link *worker_link = NULL;
```
Comment (btovar) [inline work_queue/src/work_queue_worker.c]: worker_serve_files_link

File: work_queue/src/work_queue_worker.c [hunk 11]
```diff
@@ -207,6 +209,12 @@ static int total_tasks_executed = 0;
...
+static struct link *fetch_link = NULL;
```
Comment (btovar) [inline work_queue/src/work_queue_worker.c]: worker_fetch_files_link

File: work_queue/src/work_queue_worker.c [hunk 12]
```diff
@@ -1223,6 +1238,109 @@ static int do_kill(int taskid)
 	return 1;
 }
 
+static struct link *create_worker_link(int port) {
+	struct link *worker_link = NULL;
+	worker_link = link_serve(port);
+	return worker_link;
+}
+
+static struct link *connect_to_worker(struct link *master, char *host, int port) {
```
Comment (btovar) [inline work_queue/src/work_queue_worker.c]: master not needed here.

File: work_queue/src/work_queue_worker.c [hunk 13]
```diff
@@ -1223,6 +1238,109 @@ static int do_kill(int taskid)
...
+	reset_idle_timer();
+	struct link *fetch_worker = link_connect(worker_addr,port,idle_stoptime);
+	if(!fetch_worker) {
+		return NULL;
+	}
+	return fetch_worker;
+
+}
+
+static struct link *accept_worker(struct link *master) {
```
Comment (btovar) [inline work_queue/src/work_queue_worker.c]: master not needed here.

**Comment by dthain** [comment]
Good first attempt at this capability, but it didn't keep up with the code evolution.
We will revisit this in the coming redesign of WQ.

**Pull request closed without merging**